README.md
# Indi's C++ CRC header

[![Latest release][shield-github-release]][release-latest]
[![Project license][shield-github-licence]](#licence)

Drop-in CRC calculation for C++ projects.

* Version: 0.1.0
* Project URL: https://github.com/DarkerStar/crc

## Details

This project is basically a single header - `crc.hpp` - that you can
just drop in to any C++ project to get CRC calculation.

For more information and documentation, see
[the project's wiki on Github][wiki].

## Licence

This project and all its source code is licenced under the 
[GNU General Public Licence v3.0][GPLv3].
A copy of the GPLv3 can be found in the file `LICENSE` in the project's
root directory.

## Development status

[![Project open issues][shield-github-issues]][issues]
[![Project open pull requests][shield-github-prs]][prs]

Branch              | Status
--------------------|--------
[`master`][br-mas]  | [![TravisCI for `master` branch][tcis-mas]][tcisbr-mas]
[`develop`][br-dev] | [![TravisCI for `develop` branch][tcis-dev]][tcisbr-dev]

[wiki]: https://github.com/DarkerStar/crc/wiki
[GPLv3]: https://www.gnu.org/licenses/gpl.html
[release-latest]: https://github.com/DarkerStar/crc/releases/latest
[licence]: https://github.com/DarkerStar/crc#license
[issues]: https://github.com/DarkerStar/crc/issues
[prs]: https://github.com/DarkerStar/crc/pulls
[br-mas]: https://github.com/DarkerStar/crc
[br-dev]: https://github.com/DarkerStar/crc/tree/develop

[shield-github-release]: https://img.shields.io/github/release/DarkerStar/crc.svg
[shield-github-licence]: https://img.shields.io/github/license/DarkerStar/crc.svg
[shield-github-issues]: https://img.shields.io/github/issues/DarkerStar/crc.svg
[shield-github-prs]: https://img.shields.io/github/issues-pr/DarkerStar/crc.svg
[tcis-mas]: https://travis-ci.org/DarkerStar/crc.svg?branch=master
[tcis-dev]: https://travis-ci.org/DarkerStar/crc.svg?branch=develop
[tcisbr-mas]: https://travis-ci.org/DarkerStar/crc
[tcisbr-dev]: https://travis-ci.org/DarkerStar/crc/tree/develop


CHANGELOG.md
# Change log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## 0.1.0 - 2016-09-27
### Added
- `CHANGELOG.md` file: change log.
- `LICENSE` file: project licence (contains
  [GPLv3.0](https://www.gnu.org/licenses/gpl.html)).
- `README.md` file: project information.
- `Makefile` file: root makefile, which does nothing much except running
  tests and passing test commands to the test makefile to build the
  tests.
- `indi/crc.hpp` file: the actual header file that this project is all
  about.
- `indi/crc-io.hpp` file: I/O and string formatting routines for CRCs.
- `test/Makefile` file: test makefile, which builds the tests.
- `test/test-main.cpp` file: main source code file for the tests.
- `test/calculate.cpp` file: tests for CRC calculation.
- `test/calculate-next.cpp` file: tests for calculating a single CRC.
- `test/calculate-raw.cpp` file: tests for raw CRC calculation.
- `test/crc-type.cpp` file: tests for CRC type.
- `test/generate-table.cpp` file: tests for generating lookup tables.
- `test/polynomials.cpp` file: tests for polynomial values and
  functions.
- `test/polynomials-id.cpp` file: tests for polynomial I/O and string
  formatting.
- `test/data/table-crc16` file: test lookup table values for CRC16.
- `test/data/table-crc16-ccitt` file: test lookup table values for
  CRC16-CCITT.
- `test/data/table-crc32` file: test lookup table values for CRC32.
- `test/data/table-crc32c` file: test lookup table values for CRC32c.
- `test/data/table-crc64-ecma` file: test lookup table values for
  CRC64-ECMA.

[Unreleased]: https://github.com/DarkerStar/crc/compare/v0.1.0...HEAD


test/crc-type.cpp
/* This file is part of indi-crc.
 * 
 * indi-crc is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * indi-crc is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with indi-crc.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "indi/crc.hpp"

#define BOOST_TEST_DYN_LINK
#include <boost/test/unit_test.hpp>

#include <cstdint>
#include <type_traits>

BOOST_AUTO_TEST_SUITE(crc_type_suite)

BOOST_AUTO_TEST_CASE(crc_types)
{
    BOOST_CHECK((std::is_same<std::uint_fast8_t,
        typename indi::crc::crc_type<1>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast8_t,
        typename indi::crc::crc_type<2>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast8_t,
        typename indi::crc::crc_type<3>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast8_t,
        typename indi::crc::crc_type<4>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast8_t,
        typename indi::crc::crc_type<5>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast8_t,
        typename indi::crc::crc_type<6>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast8_t,
        typename indi::crc::crc_type<7>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast8_t,
        typename indi::crc::crc_type<8>::type>::value));
    
    BOOST_CHECK((std::is_same<std::uint_fast16_t,
        typename indi::crc::crc_type<9>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast16_t,
        typename indi::crc::crc_type<10>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast16_t,
        typename indi::crc::crc_type<11>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast16_t,
        typename indi::crc::crc_type<12>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast16_t,
        typename indi::crc::crc_type<13>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast16_t,
        typename indi::crc::crc_type<14>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast16_t,
        typename indi::crc::crc_type<15>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast16_t,
        typename indi::crc::crc_type<16>::type>::value));
    
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        typename indi::crc::crc_type<17>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        typename indi::crc::crc_type<18>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        typename indi::crc::crc_type<19>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        typename indi::crc::crc_type<20>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        typename indi::crc::crc_type<21>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        typename indi::crc::crc_type<22>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        typename indi::crc::crc_type<23>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        typename indi::crc::crc_type<24>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        typename indi::crc::crc_type<25>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        typename indi::crc::crc_type<26>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        typename indi::crc::crc_type<27>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        typename indi::crc::crc_type<28>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        typename indi::crc::crc_type<29>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        typename indi::crc::crc_type<30>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        typename indi::crc::crc_type<31>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        typename indi::crc::crc_type<32>::type>::value));
    
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<33>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<34>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<35>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<36>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<37>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<38>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<39>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<40>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<41>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<42>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<43>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<44>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<45>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<46>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<47>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<48>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<49>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<50>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<51>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<52>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<53>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<54>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<55>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<56>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<57>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<58>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<59>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<60>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<61>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<62>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<63>::type>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        typename indi::crc::crc_type<64>::type>::value));
}

BOOST_AUTO_TEST_CASE(crc_type_ts)
{
    BOOST_CHECK((std::is_same<std::uint_fast8_t,
        indi::crc::crc_type_t<1>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast8_t,
        indi::crc::crc_type_t<2>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast8_t,
        indi::crc::crc_type_t<3>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast8_t,
        indi::crc::crc_type_t<4>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast8_t,
        indi::crc::crc_type_t<5>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast8_t,
        indi::crc::crc_type_t<6>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast8_t,
        indi::crc::crc_type_t<7>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast8_t,
        indi::crc::crc_type_t<8>>::value));
    
    BOOST_CHECK((std::is_same<std::uint_fast16_t,
        indi::crc::crc_type_t<9>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast16_t,
        indi::crc::crc_type_t<10>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast16_t,
        indi::crc::crc_type_t<11>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast16_t,
        indi::crc::crc_type_t<12>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast16_t,
        indi::crc::crc_type_t<13>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast16_t,
        indi::crc::crc_type_t<14>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast16_t,
        indi::crc::crc_type_t<15>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast16_t,
        indi::crc::crc_type_t<16>>::value));
    
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        indi::crc::crc_type_t<17>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        indi::crc::crc_type_t<18>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        indi::crc::crc_type_t<19>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        indi::crc::crc_type_t<20>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        indi::crc::crc_type_t<21>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        indi::crc::crc_type_t<22>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        indi::crc::crc_type_t<23>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        indi::crc::crc_type_t<24>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        indi::crc::crc_type_t<25>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        indi::crc::crc_type_t<26>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        indi::crc::crc_type_t<27>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        indi::crc::crc_type_t<28>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        indi::crc::crc_type_t<29>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        indi::crc::crc_type_t<30>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        indi::crc::crc_type_t<31>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast32_t,
        indi::crc::crc_type_t<32>>::value));
    
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<33>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<34>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<35>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<36>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<37>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<38>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<39>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<40>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<41>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<42>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<43>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<44>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<45>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<46>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<47>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<48>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<49>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<50>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<51>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<52>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<53>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<54>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<55>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<56>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<57>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<58>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<59>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<60>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<61>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<62>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<63>>::value));
    BOOST_CHECK((std::is_same<std::uint_fast64_t,
        indi::crc::crc_type_t<64>>::value));
}

BOOST_AUTO_TEST_SUITE_END()

test/test-main.cpp
/* This file is part of indi-crc.
 * 
 * indi-crc is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * indi-crc is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with indi-crc.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "indi/crc.hpp"

#define BOOST_TEST_DYN_LINK
#define BOOST_TEST_MODULE crc
#include <boost/test/unit_test.hpp>


indi/crc-io.hpp
/* This file is part of indi-crc.
 * 
 * indi-crc is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * indi-crc is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with indi-crc.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef INDI_INC_CRC_IO_
#define INDI_INC_CRC_IO_

#include <locale>
#include <sstream>
#include <string>
#include <type_traits>

#include "indi/crc.hpp"

namespace indi {
namespace crc {

namespace polynomials {

//! Convert a CRC polynomial from normal encoded form to a string
//! representation.
//! 
//! CRC polynomials are normally encoded as an unsigned integer by
//! having the each bit represent the coefficient for the term whose
//! power is equal to the bit position. So if the value of bit `N` is
//! `b`, this represents the term in the polynomial `b * x^N`.
//! 
//! (The `^` character represents exponentiation operators. So `x^2` is
//! meant to be interpreted as "x squared", and `x^15` is "x raised to
//! the power of 15".)
//! 
//! In addition, for a `M`-bit CRC, the coefficient for `x^M` is not
//! encoded, and is implicitly 1.
//! 
//! So for  the 16-bit CRC polynomial `0x8005u`, bits 15, 2, and 0 are
//! set, so this function will convert it into the string:
//! 
//!     x^16 + x^15 + x^2 + 1
//! 
//! The `Bits` template parameter sets the CRC size. Only the least
//! significant `Bits` bits are considered - any higher bits are
//! ignored. So treating `0x50108005uL` as a 16-bit CRC polynomial will
//! give the same result as above. But treating it as a 32-bit
//! polynomial would result in:
//! 
//!     x^32 + x^30 + x^28 + x^15 + x^2 + 1
//! 
//! The format of the result string is guaranteed to be as follows:
//! 
//! 1.  For an `N`-bit CRC polynomial, the first term is always `x^N`.
//! 2.  For all bits that are set to 1 between bit `N` and bit 1
//!     (where bit 0 is the least significant bit), not inclusive,
//!     a term of the form ` + x^B` is appended, where `B` is the
//!     bit number.
//! 3.  If bit 1 is set, ` + x` is appended.
//! 4.  If bit 0 is set (which it always should be, for a CRC
//!     polynomial encoded in this form), ` + 1` is appended.
//! 
//! Note that numbers are always formatted according to the default "C"
//! locale, no matter what the current global locale is.
//! 
//! \tparam Bits  The CRC bit-size.
//! 
//! \tparam T The type of the encoded polynomial value.
//! 
//! No checking is done to ensure that `T` has at least `Bits` bits.
//! If `T` is too small to hold `Bits` bits, all bits missing are
//! presumed to be 0.
//! 
//! \param polynomial  The encoded polynomial value.
//! 
//! No checking is done to ensure that it is a valid encoded CRC
//! polynomial.
//! 
//! Only the least significant `Bits` bits are used.
//! 
//! \returns A string with the CRC polynomial encoded as described in
//!          the details.
//! 
template <std::size_t Bits, typename T>
auto to_string(T polynomial)
{
	static_assert(std::is_integral<T>::value,
        "CRC type must be integer");
	static_assert(std::is_unsigned<T>::value,
        "CRC type must be unsigned");
	static_assert(Bits > 0, "0-bit CRCs make no sense");
    
    // TODO: Handle the (rare) special case where Bits == 1.
    //       This would be a perfect job for constexpr if.
    
    auto oss = std::ostringstream{};
    
    // Explicitly imbue the "C" locale, just in case the global locale
    // has been set to something wacky.
    oss.imbue(std::locale::classic());
    
    // Handle the implicit order term.
    oss << "x^" << Bits;
    
    // Handle all exponents except 0 and 1.
    for (auto bit = Bits; bit > std::size_t{2}; --bit)
    {
        // IMPORTANT: The "1" must be cast to T before shifting.
        if (polynomial & (T(1u) << (bit - 1)))
            oss << " + x^" << (bit - 1);
    }
    
    // Handle "x" term.
    if (Bits > std::size_t{1} && (polynomial & 0x2u))
        oss << " + x";
    
    // Handle constant term.
    // Bits can never be 0, so there's no need to check. Technically
    // the constant term should always be 1 for a valid CRC polynomial,
    // so this if is really not necessary, but meh. It costs basically
    // nothing, and it makes the function slightly more generic.
    if (polynomial & 0x1u)
        oss << " + 1";
    
    return oss.str();
}

} // namespace polynomials

} // namespace crc
} // namespace indi

#endif // include guard
